pkg/insight/insightstore/deploymentstore.go

```diff
@@ -17,26 +17,361 @@ package insightstore
 import (
 	"context"
 	"errors"
+	"fmt"
+	"time"
 
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/pipe-cd/pipecd/pkg/filestore"
+	"github.com/pipe-cd/pipecd/pkg/insight"
 	"github.com/pipe-cd/pipecd/pkg/model"
 )
 
 var (
-	errUnimplemented = errors.New("unimplemented")
+	errInvalidArg    = errors.New("invalid arg")
+	errLargeDuration = errors.New("too large duration")
+	errExceedMaxSize = errors.New("exceed max file size")
+	errInconsistent  = errors.New("no consistency between meta and chunk")
+)
+
+const (
+	maxChunkByteSize = 1 * 1024 * 1024 // 1MB
+	metaFileName     = "meta.json"
+
+	maxDuration time.Duration = 24 * 365 * 2 * time.Hour // 2 year
 )
 
 type DeploymentStore interface {
-	// List returns slice of Deployment sorted by startedAt ASC.
-	List(ctx context.Context, projectID string, from, to int64, minimumVersion *model.InsightDeploymentVersion) ([]*model.InsightDeployment, error)
+	// List returns slice of Deployment sorted by CompletedAt ASC.
+	List(ctx context.Context, projectID string, from, to int64, minimumVersion model.InsightDeploymentVersion) ([]*model.InsightDeployment, error)
+
+	Put(ctx context.Context, projectID string, deployments []*model.InsightDeployment, version model.InsightDeploymentVersion) error
+}
+
+// List returns slice of Deployment sorted by CompletedAt ASC.
+func (s *store) List(ctx context.Context, projectID string, from, to int64, minimumVersion model.InsightDeploymentVersion) ([]*model.InsightDeployment, error) {
+	fromTime := time.Unix(from, 0)
+	fromYear := fromTime.Year()
+	toTime := time.Unix(to, 0)
+	toYear := toTime.Year()
+
+	sub := toTime.Sub(fromTime)
+	if sub < 0 {
+		return nil, errInvalidArg
+	}
+	if sub > maxDuration {
+		return nil, errLargeDuration
+	}
+
+	var result []*model.InsightDeployment
+	for year := fromYear; year <= toYear; year++ {
+		dirPath := determineDeploymentDirPath(year, projectID)
+		meta := model.InsightDeploymentChunkMetadata{}
+		err := s.loadDataFromFilestore(ctx, fmt.Sprintf("%s/%s", dirPath, metaFileName), &meta)
+		if errors.Is(err, filestore.ErrNotFound) {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		keys := (&meta).ListChunkNames(from, to)
+
+		for _, key := range keys {
+			data := model.InsightDeploymentChunk{}
+			err := s.loadDataFromFilestore(ctx, fmt.Sprintf("%s/%s", dirPath, key), &data)
+			if err != nil {
+				return nil, err
+			}
+
+			// Maybe we should use metadata instead of loading chunkdata
+			if data.Version >= minimumVersion {
+				deployments := (&data).ExtractDeployments(from, to)
+				result = append(result, deployments...)
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// deployments must be sorted by CompletedAt,
```

Comment:
should make this comment/not as Put interface description at L48 in order to available to see from the caller side.